File: PI_web/farmagrid/src/services/authService.js
import { apiFetch, setToken } from './api.js';

// ─────────────────────────────────────────────────────────────────────────
// ─────────────────────────────────────────────────────────────────────────
export const TIPO_LOGIN = { paciente: 3, profissional: 1 };

export const PERFIL_API_PARA_FRONT = {
  paciente: 'paciente',
  medico: 'profissional',
  farmacia: 'farmacia',
  balconista: 'farmacia',
  caixa: 'farmacia',
};

export function mapearPerfilFront(perfilApi) {
  return PERFIL_API_PARA_FRONT[perfilApi] || 'paciente';
}

export async function login(email, senha, tipoFront) {
  const tipo = TIPO_LOGIN[tipoFront] ?? TIPO_LOGIN.paciente;
  const res = await apiFetch('/auth/login', {
    method: 'POST',
    body: JSON.stringify({ email, senha, tipo }),
  });
  setToken(res.token);
  return res;
}

// Endpoints públicos de cadastro.
export async function cadastroPaciente(dados) {
  return apiFetch('/auth/cadastro/paciente', { method: 'POST', body: JSON.stringify(dados) });
}

export async function cadastroMedico(dados) {
  return apiFetch('/auth/cadastro/medico', { method: 'POST', body: JSON.stringify(dados) });
}

export async function cadastroFarmacia(dados) {
  const cnpj = String(dados.cnpj || '').replace(/\D/g, '');
  const cadastros = JSON.parse(localStorage.getItem('farmagrid_farmacias_demo') || '[]');

  if (cadastros.some(farmacia => farmacia.cnpj === cnpj)) {
    throw new Error('CNPJ já cadastrado nesta demonstração.');
  }

  const farmacia = {
    id: Date.now(),
    nome: dados.nome,
    cnpj,
    cep: String(dados.cep || '').replace(/\D/g, ''),
    telefone: String(dados.telefone || '').replace(/\D/g, ''),
    idCidade: dados.idCidade,
    fotoPerfil: dados.fotoPerfil || null,
    email: dados.email,
  };

  localStorage.setItem('farmagrid_farmacias_demo', JSON.stringify([...cadastros, farmacia]));
  return farmacia;
}

// GET/PUT /api/pacientes/{id}/config (PerfilController) — substitui o
// inexistente /auth/me. Retorna nome, cpf, idade, endereço completo, plano,
// foto (byte[]) e status de assinatura.
export async function buscarConfigPaciente(idPaciente) {
  return apiFetch(`/api/pacientes/${idPaciente}/config`);
}

export async function atualizarConfigPaciente(idPaciente, payload) {
  return apiFetch(`/api/pacientes/${idPaciente}/config`, {
    method: 'PUT',
    body: JSON.stringify(payload),
  });
}

// PUT /api/pacientes/{id}/foto — espera { foto: byte[] } (Jackson serializa
// byte[] como string base64 automaticamente), por isso mandamos só a parte
// base64 da data URL (sem o prefixo "data:image/...;base64,").
export async function salvarFotoPaciente(idPaciente, base64SemPrefixo) {
  return apiFetch(`/api/pacientes/${idPaciente}/foto`, {
    method: 'PUT',
    body: JSON.stringify({ foto: base64SemPrefixo }),
  });
}

// PUT /api/logins/{id}/senha — id aqui é o id da linha `login`, não o
// idPaciente. Ainda não há campo de troca de senha na tela de Editar Perfil.
export async function alterarSenha(idLogin, senhaAtual, novaSenha) {
  return apiFetch(`/api/logins/${idLogin}/senha`, {
    method: 'PUT',
    body: JSON.stringify({ senhaAtual, novaSenha }),
  });
}

export function logout() {
  setToken(null);
}

// Monta o payload de CadastroPacienteRequest / CadastroMedicoRequest
// (com os nomes de campo exatos que a API espera).
export function montarCadastro(form, tipoFront) {
  if (tipoFront === 'profissional') {
    return {
      email: form.email,
      senha: form.senha,
      nome: form.nome,
      crm: form.crm || '00000-SP',
      especialidade: form.especialidade || 'Clínica Geral',
      clinica: form.clinica || 'FarmaGrid',
      enderecoClinica: form.enderecoClinica || '',
      telefone: form.telefone || '',
      endereco: form.endereco || '',
      dataNascimento: form.dataNascimento || null,
      rqe: form.rqe || '',
      subespecialidades: form.subespecialidades || '',
      horarioInicio: form.horarioInicio || '08:00',
      horarioTermino: form.horarioTermino || '18:00',
      tempoConsulta: form.tempoConsulta || '30',
      valorConsulta: form.valorConsulta || '0',
      tipoAtendimento: form.tipoAtendimento || 'Presencial',
    };
  }

  // paciente
  return {
    email: form.email,
    senha: form.senha,
    nome: form.nome,
    cpf: form.cpf,
    dataNascimento: form.dataNascimento || null,
    sexo: form.sexo || 'M',
    rua: form.rua || '',
    numCasa: form.numCasa ? Number(form.numCasa) : 0,
    bairro: form.bairro || '',
    cidade: form.cidade || '',
    estado: form.estado || '',
    idCidade: null, // ⚠️ não há seletor de cidade (tabela de cidades) no formulário ainda
    idPlano: null,  // ⚠️ não há seletor de plano de saúde no formulário ainda
    telefone: form.telefone || '',
    cep: form.cep || '',
    tipoSanguineo: form.tipoSanguineo || '',
    contatoEmergenciaNome: form.contatoEmergenciaNome || '',
    contatoEmergenciaTelefone: form.contatoEmergenciaTelefone || '',
  };
}

File: FarmaGrid+_MOBILE/lib/services/auth_service.dart
import 'dart:convert';

import 'package:http/http.dart' as http;
import 'package:flutter_secure_storage/flutter_secure_storage.dart';

import '../config/api_config.dart';
import '../models/usuario_logado.dart';

class TipoLogin {
  static const int medico = 1;
  static const int paciente = 3;
}

class ApiException implements Exception {
  final String mensagem;

  ApiException(this.mensagem);

  @override
  String toString() => mensagem;
}

class AuthService {
  static const FlutterSecureStorage _storage = FlutterSecureStorage();

  static const String _chaveSessao = 'sessao_usuario';

  static UsuarioLogado? usuarioLogado;

  static Future<UsuarioLogado> login(
    String email,
    String senha,
    int tipo,
  ) async {
    try {
      final resposta = await http.post(
        Uri.parse('${ApiConfig.baseUrl}/auth/login'),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({'email': email, 'senha': senha, 'tipo': tipo}),
      );

      if (resposta.statusCode != 200) {
        throw ApiException(_extrairMensagemErro(resposta));
      }

      final dynamic json = jsonDecode(resposta.body);

      if (json is! Map<String, dynamic>) {
        throw ApiException('Resposta inválida recebida do servidor.');
      }

      final usuario = UsuarioLogado.fromJson(json);

      usuarioLogado = usuario;

      await _storage.write(
        key: _chaveSessao,
        value: jsonEncode(usuario.toJson()),
      );

      return usuario;
    } on ApiException {
      rethrow;
    } catch (e) {
      throw ApiException(
        'Não foi possível conectar ao servidor. Verifique se a API está em execução em ${ApiConfig.baseUrl}.',
      );
    }
  }

  static Future<void> cadastrarPaciente({
    required String email,
    required String senha,
    required String nome,
    required String cpf,
    required String dataNascimento,
    required String sexo,
    required String rua,
    required int? numCasa,
    required String bairro,
    required String cidade,
    required String estado,
    required String telefone,
    required String cep,
    required String tipoSanguineo,
    required String contatoEmergenciaNome,
    required String contatoEmergenciaTelefone,
  }) async {
    await _post('/auth/cadastro/paciente', {
      'email': email,
      'senha': senha,
      'nome': nome,
      'cpf': cpf.replaceAll(RegExp(r'[^0-9]'), ''),
      'dataNascimento': dataNascimento,
      'sexo': sexo,
      'rua': rua,
      'numCasa': numCasa,
      'bairro': bairro,
      'cidade': cidade,
      'estado': estado,
      'telefone': telefone,
      'cep': cep,
      'tipoSanguineo': tipoSanguineo,
      'contatoEmergenciaNome': contatoEmergenciaNome,
      'contatoEmergenciaTelefone': contatoEmergenciaTelefone,
    });
  }

  static Future<void> cadastrarMedico({
    required String email,
    required String senha,
    required String nome,
    required String crm,
    required String especialidade,
    required String clinica,
    required String enderecoClinica,
    required String telefone,
    required String endereco,
    required String dataNascimento,
    required String rqe,
    required String subespecialidades,
    required String horarioInicio,
    required String horarioTermino,
    required String tempoConsulta,
    required String valorConsulta,
    required String tipoAtendimento,
  }) async {
    await _post('/auth/cadastro/medico', {
      'email': email,
      'senha': senha,
      'nome': nome,
      'crm': crm,
      'especialidade': especialidade,
      'clinica': clinica,
      'enderecoClinica': enderecoClinica,
      'telefone': telefone,
      'endereco': endereco,
      'dataNascimento': dataNascimento,
      'rqe': rqe,
      'subespecialidades': subespecialidades,
      'horarioInicio': horarioInicio,
      'horarioTermino': horarioTermino,
      'tempoConsulta': tempoConsulta,
      'valorConsulta': valorConsulta,
      'tipoAtendimento': tipoAtendimento,
    });
  }

  static Future<void> _post(String path, Map<String, dynamic> body) async {
    try {
      final resposta = await http.post(
        Uri.parse('${ApiConfig.baseUrl}$path'),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode(body),
      );

      if (resposta.statusCode < 200 || resposta.statusCode >= 300) {
        throw ApiException(_extrairMensagemErro(resposta));
      }
    } on ApiException {
      rethrow;
    } catch (e) {
      throw ApiException(
        'Não foi possível conectar ao servidor. Verifique se a API está em execução em ${ApiConfig.baseUrl}.',
      );
    }
  }

  static Future<bool> restaurarSessao() async {
    final salvo = await _storage.read(key: _chaveSessao);

    if (salvo == null) {
      usuarioLogado = null;
      return false;
    }

    try {
      final dynamic json = jsonDecode(salvo);

      if (json is! Map<String, dynamic>) {
        throw Exception('Sessão inválida');
      }

      usuarioLogado = UsuarioLogado.fromJson(json);

      return true;
    } catch (_) {
      await _storage.delete(key: _chaveSessao);

      usuarioLogado = null;

      return false;
    }
  }

  static Future<void> logout() async {
    await _storage.delete(key: _chaveSessao);

    usuarioLogado = null;
  }

  static String _extrairMensagemErro(http.Response resposta) {
    try {
      final dynamic corpo = jsonDecode(resposta.body);

      if (corpo is Map) {
        final campoMensagem =
            corpo['message'] ??
            corpo['mensagem'] ??
            corpo['error'] ??
            corpo['erro'];
        if (campoMensagem != null) {
          return campoMensagem.toString();
        }
      }

      if (corpo is String && corpo.trim().isNotEmpty) {
        return corpo;
      }

      return resposta.body.isNotEmpty
          ? resposta.body
          : 'Erro inesperado ao conectar com o servidor.';
    } catch (_) {
      if (resposta.body.isNotEmpty) {
        return resposta.body;
      }

      return 'Erro inesperado ao conectar com o servidor.';
    }
  }
}
